src/dftbp/dftb/periodic.F90

```diff
@@ -805,39 +804,27 @@ subroutine allocateNeighbourArrays(neigh, maxNeighbour, nAtom, env)
     !> Environment settings
     type(TEnvironment), intent(in), optional :: env
 
-  #:if WITH_MPI
     integer :: dataLength
-  #:endif
+
+    dataLength = (maxNeighbour + 1) * nAtom
 
     if (neigh%useMpiWindows_) then
     #:if WITH_MPI
-      if (associated(neigh%iNeighbourMem_)) then
-        call neigh%iNeighbourWin_%free()
-      end if
-      if (associated(neigh%neighDist2Mem_)) then
-        call neigh%neighDist2Win_%free()
-      end if
-
-      dataLength = (maxNeighbour + 1) * nAtom
-
+      if (associated(neigh%iNeighbourMem_)) call neigh%iNeighbourWin_%free()
       call neigh%iNeighbourWin_%allocate_shared(env%mpi%nodeComm, dataLength, neigh%iNeighbourMem_)
+      if (associated(neigh%neighDist2Mem_)) call neigh%neighDist2Win_%free()
       call neigh%neighDist2Win_%allocate_shared(env%mpi%nodeComm, dataLength, neigh%neighDist2Mem_)
-
-      neigh%iNeighbour(0:maxNeighbour, 1:nAtom) => neigh%iNeighbourMem_(1:dataLength)
-      neigh%neighDist2(0:maxNeighbour, 1:nAtom) => neigh%neighDist2Mem_(1:dataLength)
     #:endif
     else
-      if (associated(neigh%iNeighbour)) then
-        deallocate(neigh%iNeighbour)
-      end if
-      if (associated(neigh%neighDist2)) then
-        deallocate(neigh%neighDist2)
-      end if
-
-      allocate(neigh%iNeighbour(0:maxNeighbour,1:nAtom))
-      allocate(neigh%neighDist2(0:maxNeighbour,1:nAtom))
+      if (associated(neigh%iNeighbourMem_)) deallocate(neigh%iNeighbourMem_)
+      allocate(neigh%iNeighbourMem_(dataLength))
+      if (associated(neigh%neighDist2Mem_)) deallocate(neigh%neighDist2Mem_)
+      allocate(neigh%neighDist2Mem_(dataLength))
     end if
 
+    neigh%iNeighbour(0:maxNeighbour, 1:nAtom) => neigh%iNeighbourMem_(1:dataLength)
```

Comment:
For the serial case, would it be more efficient to move_alloc and drop the allocation?

Reply:
I'd like to allocate the same variable in both (win, non-win) cases, otherwise the code is too confusing and error-prone as we have seen... This way, it would always consistently be the two `*Mem` fields, which must be monitored/handled, if allocation/deallocation happens. I hope, this leads to less confusion and bugs.